### PR TITLE
Sylvia whittle/298 improve flattening

### PR DESCRIPTION
**Pull request by SylviaWhittle**
This Pull Request improves the flattening algorithms used in the filtering stage of processing. It also adds support for differing upper and lower standard deviation thresholds in the filtering and grain finding stages. 

- Update and improve the median flattening function (now `median_flattening`, previously `align_rows`) to be faster, also fix a few bugs where edge cases would cause rows not to be aligned properly.

- Update the plane flattening (`initial_tilt_removal`) to use lines of best fit of median samples for each row and column rather than the median from the initial and final row and columns, resulting in accurate plane tilt removal rather than the previous approximation used. 

- Update the masking functions, simplifying them and slightly increasing performance.

- Add `quadratic_removal`. This method is similar to the plane flattening, however instead of a linear fit, this uses a quadratic fit. This should remove most of the quadratic bowing that some AFM scans can exhibit when scanning over a large area.

- Added height-frequency histogram outputs for each image produced in the processing of an image. These plots provide useful insight into the distribution of heights within the image. This can be very useful in determining the quality of the flattening too. In the future I plan on adding an automated tool to determine the quality of the flattening using such plots.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -36,6 +36,7 @@ tests =
   pytest-cov
   pytest-mpl
   pytest-regtest
+  py
```
Comment (ns-rse) [inline setup.cfg:39]: Minor but its useful to list package dependencies in alphabetical order for ease of reading, however in this instance because you merged `dev` you've now got `py` in there twice so this line can be removed.

File: .gitignore [hunk 2]
```diff
@@ -46,3 +46,5 @@ _build/
 
 # Tests and docs
 .coverage
+
+!Flattening development/images/*.png
```
Comment (ns-rse) [inline .gitignore]: Useful to have this ignored locally whilst developing but `.gitignore` should be general and work for everyone and most won't have this.

Simply not staging the file would be fine in future when you add folders during develpoment.

File: tests/_regtest_outputs/test_run_topostats.test_process_scan_lower.out [hunk 3]
```diff
@@ -1,3 +1,2 @@
...
-0                312.31095  135.689968     2.04927    6.596742      4.42801       4.502005   -0.838749   -0.473677      -0.602617    -0.607314 -31.864748  52.468362           106.156919                 5.459192                  12.95475               70.722472      2.373016     lower  12.966989   5.459192        44.837437     False             5.948583  minicircle.spm
+   Molecule Number  centre_x  centre_y  radius_min  radius_max  radius_mean  radius_median  height_min  height_max  height_median  height_mean  volume  area  area_cartesian_bbox  smallest_bounding_width  smallest_bounding_length  smallest_bounding_area  aspect_ratio  Contour Lengths  Circular  End to End Distance  Image Name
+0              NaN       NaN       NaN         NaN         NaN          NaN            NaN         NaN         NaN            NaN          NaN     NaN   NaN                  NaN                      NaN                       NaN                     NaN           NaN              NaN       NaN                  NaN         NaN
```
Comment (ns-rse) [inline tests/_regtest_outputs/test_run_topostats.test_process_scan_lower.out:2]: See above comment against `tests/_regtest_outputs/test_run_topostats.test_process_scan_both.out` about no longer having a test for lower scan threshold as nothing is detected.

File: tests/conftest.py [hunk 4]
```diff
@@ -218,22 +223,22 @@ def test_filters_random_with_mask(filter_config: dict, test_filters: Filters, im
...
         test_filters_random_with_mask.images["initial_tilt_removal"], mask=test_filters_random_with_mask.images["mask"]
     )
     test_filters_random_with_mask.images["masked_tilt_removal"] = test_filters_random_with_mask.remove_tilt(
         test_filters_random_with_mask.images["masked_align"], mask=test_filters_random_with_mask.images["mask"]
     )
 
-    test_filters_random_with_mask.images["zero_averaged_background"] = test_filters_random_with_mask.average_background(
-        test_filters_random_with_mask.images["masked_tilt_removal"], mask=test_filters_random_with_mask.images["mask"]
-    )
+    # test_filters_random_with_mask.images["zero_averaged_background"] = test_filters_random_with_mask.average_background(
```
Comment (ns-rse) [inline tests/conftest.py]: Commented lines that are no longer used can be removed. I personally find it makes the diffs in PRs easier to read/review.

File: tests/conftest.py [hunk 5]
```diff
@@ -377,23 +393,34 @@ def minicircle_masked_tilt_removal(minicircle_masked_align: Filters) -> Filters:
...
-    ] = minicircle_masked_tilt_removal.average_background(
+def minicircle_masked_quadratic_removal(minicircle_masked_tilt_removal: Filters) -> Filters:
+    """Secondary quadratic removal using mask."""
+    minicircle_masked_tilt_removal.images["masked_quadratic_removal"] = minicircle_masked_tilt_removal.remove_quadratic(
         minicircle_masked_tilt_removal.images["masked_tilt_removal"], mask=minicircle_masked_tilt_removal.images["mask"]
     )
     return minicircle_masked_tilt_removal
 
 
+# @pytest.fixture
```
Comment (ns-rse) [inline tests/conftest.py]: Commented sections can be removed.

File: .gitignore (hunk 2 above)
Comment (SylviaWhittle) [reply .gitignore]: I think that this was removed already as of [2737cf8](https://github.com/AFM-SPM/TopoStats/pull/351/commits/2737cf8c9741ddade9651b3f2f9f9d264ec00e60). Sorry if I've got this wrong.

File: tests/conftest.py (hunk 4 above)
Comment (SylviaWhittle) [reply tests/conftest.py]: I believe I fixed this earlier too

File: tests/_regtest_outputs/test_run_topostats.test_process_scan_both.out [hunk 6]
```diff
@@ -1,24 +1,23 @@
...
+                threshold    centre_x    centre_y  radius_min  radius_max  radius_mean  radius_median  height_min  height_max  height_median  height_mean       volume        area  area_cartesian_bbox  smallest_bounding_width  smallest_bounding_length  smallest_bounding_area  aspect_ratio  max_feret  min_feret      Image Name  Contour Lengths Circular  End to End Distance
+Molecule Number                                                                                                                                                                                                                                                                                                                                                                      
+0                   upper   18.739027   96.544583    5.914298   20.891859    15.564967      16.245908    0.776713    3.922363       2.093704     1.999236  1274.860013  637.673631          1137.465287                24.428348                 40.849626              997.888853      1.672222  41.334203  24.232548  minicircle.spm        46.508479     True             0.000000
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/_regtest_outputs/test_run_topostats.test_process_scan_both.out]: I have changed the `threshold_std_dev_lower` value in the `grains` section of `default_config.yaml` to be `1.0`. This should cause a grain to be detected, and therefore fix the issue you raised there. A potential problem comes with the fact that I cannot change the value in the `filters` section of the code, because if I do, then this will make the flattening worse. The value for the flattening section is `10.0` as to ensure that no data will be detected for lower parts by default, since by default we want it to be configured for AFM images where we only want to see and are expecting to see high up grains. 

Perhaps I should make a `direction` parameter like with `grains` to enable / disable upper / lower threshold values instead of setting the default lower one to filter out 99.9999..% of the data.

File: utils/heightthresholding.py [hunk 7]
```diff
@@ -62,7 +62,7 @@ def otsuthresholdgrainfinding(data, k):
     # # Invert mask to maks things below the membrane
     mask.grains_invert()
 
-    gwy.gwy_process_func_run("align_rows", data, gwy.RUN_IMMEDIATE)
```
Comment (MaxGamill-Sheffield) [inline utils/heightthresholding.py:65]: If this is a Gwyddion function (leftover code I know), have you added the "median_flatten" here as a reference to what it _should_ be doing when equated to our code?

File: topostats/plotting_dictionary.yaml [hunk 8]
```diff
@@ -45,9 +50,9 @@ masked_tilt_removal:
   title: "Secondary Tilt Removal (Masked)"
   image_type: "non-binary"
   core_set: false
-zero_averaged_background:
-  filename: "07-zero_average_background"
-  title: "Zero Average Background"
+masked_quadratic_removal:
+  filename: "xx-quadratic_removal_masked"
```
[the remark this answers is not in the document]
Comment (MaxGamill-Sheffield) [reply topostats/plotting_dictionary.yaml:54]: Also if you're here and its not too much effort, the binary image filename still has ".png" in it so ends up producing an image named "<fname>.png.png", if you can remove the .png bit?

**Review by MaxGamill-Sheffield — CHANGES_REQUESTED** [review]
Looks pretty good, just a few things to tidy-up I think such as commented tests and xx images.

The only other thing I'd mention is the documentation (`docs/<something>.md` files). Are we going to add the workflow changes in there at some point?

File: tests/tracing/_regtest_outputs/test_dnatracing.test_tracestats.out [hunk 9]
```diff
@@ -1,24 +1,24 @@
                  Contour Lengths  Circular  End to End Distance Image Name Basename
-Molecule Number                                                                    
-0                     225.487215     False            35.015017        tmp        .
+Molecule Number
+0                     225.607081     False            35.015017        tmp        .
 1                      45.824632      True             0.000000        tmp        .
-2                     110.326409     False            21.881598        tmp        .
-3                      94.436535      True             0.000000        tmp        .
+2                     127.758534     False            22.092483        tmp        .
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/tracing/_regtest_outputs/test_dnatracing.test_tracestats.out:5]: I have been looking into this but am struggling to find a good way of visualising it. @ns-rse do you have any suggestions as to how I should investigate this change? I tried visualising it by running topostats with the parameters used in the test on both my branch and dev, however I get a problem where my branch doesn't pick up on one of the grains, indicating that there is something amiss in my configuration when trying to imitate the tests. 

Perhaps there is a way to get the regression test to save an image of the grain masks during the test?

File: topostats/filters.py [hunk 10]
```diff
@@ -86,90 +89,97 @@ def __init__(
...
         if mask is not None:
-            image = np.ma.masked_array(image, mask=mask, fill_value=np.nan).filled()
-            LOGGER.info(f"[{self.filename}] : Masking enabled")
+            read_matrix = np.ma.masked_array(image, mask=mask, fill_value=np.nan).filled()
+            LOGGER.info("plane removal with mask")
+        else:
+            read_matrix = image
+            LOGGER.info("plane removal without mask")
+
+        # LOBF
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/filters.py:118]: LIne of best fit, I'll make that more clear though

File: utils/heightthresholding.py (hunk 7 above)
Comment (SylviaWhittle) [reply utils/heightthresholding.py:65]: Shouldn't this be deleted if it's calling gwy, since AFAIK gwy is no longer able to be ran / loaded?

File: topostats/plotting_dictionary.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/plotting_dictionary.yaml:54]: Thank you!! When I added the xx's when testing I KNEW I would forget... and I did! 😆 

I'll rename the binary file name too 👍

File: utils/heightthresholding.py (hunk 7 above)
Comment (ns-rse) [reply utils/heightthresholding.py:65]: This is an old file as its under `utils/` I think I made a comment on this too and we should definitely purge all these old things before releasing.

File: tests/_regtest_outputs/test_run_topostats.test_process_scan_both.out (hunk 6 above)
[the remark this answers is not in the document]
Comment (ns-rse) [reply tests/_regtest_outputs/test_run_topostats.test_process_scan_both.out]: Completely understand going with the default and what is most likely to be used.

One of the aspects of testing though is to try and capture edge cases. The lower threshold was introduced for a reason, my memory is hazy but perhaps something to do with scans of silicone, and so we should have tests that cover that too.

Modifying parameters in tests is what we should do to cover this and its fairly straight-forward to setup fixtures of configs with modified values. The `process_scan_config()` fixture defined on lines 44-53 of `tests/conftest.py` shows how to do this.

However, if `Filters()` will _never_ use a lower threshold then there is no need to even have a `threshold_std_dev_lower` in there to be tested.

File: tests/tracing/_regtest_outputs/test_dnatracing.test_tracestats.out (hunk 9 above)
[the remark this answers is not in the document]
Comment (ns-rse) [reply tests/tracing/_regtest_outputs/test_dnatracing.test_tracestats.out:5]: A common plot that I use when comparing two methods is [Bland-Altman plot](https://en.wikipedia.org/wiki/Bland–Altman_plot) and so you could for each metric plot the "old" v "new" values and see if there are any extreme values.

That doesn't tell you what has changed though and the only way I can think of doing that is to compare current `dev` output (i.e. before any change to flattening) to this branch. Grains are numbered from top-left to bottom right if I remember correctly (or perhaps the reverse :thinking: ) and what would really be required is to extract plots of the skeletonised molecules and compare them as my understanding is that these statistics are based on the post-skeletonisation traces.

Not sure if that is helpful (pretty sure the Bland-Altman stuff isn't immediately relevant but worth bearing in mind when comparing analytical methods in general).

**Review by ns-rse — CHANGES_REQUESTED** [review]
Great work @SylviaWhittle most of my comments are pretty minor really, pedantic things like docstrings, formatting of logging info.

I think it would be really useful to have a descriptive documentation of what is being done with regards to flattening beyond the current Mermaid diagram (same is true for many other stages).

Shout if there is anything I can help with.

**Comment by MaxGamill-Sheffield** [comment]
Great work Sylvia!

I also forgot to mention thats the histograms are still produced when the `image_set: core`. I think these should only be made when `image_set: all` and also reside in the `<filename>` folder.

File: topostats/utils.py [hunk 11]
```diff
@@ -204,7 +205,7 @@ def get_thresholds(
     image: np.ndarray,
     threshold_method: str,
     otsu_threshold_multiplier: float = None,
-    threshold_std_dev: float = None,
+    threshold_std_dev: tuple = None,
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/utils.py:208]: Do you mean change it in the config file so that instead of:

```
threshold_std_dev_upper: 1.0
threshold_std_dev_lower: 1.0
```

we have 

```
threshold_std_dev: [1.0, 1.0]
```

if so, should this change also be made for `threshold_absolute_upper` and `threshold_absolute_lower`? 

I think that it might be better to leave the values in the config file as separate lines to maintain simplicity (not having to figure out which is upper and lower) though I'm very happy to have pushback on this - just a very minor preference.

File: utils/heightthresholding.py (hunk 7 above)
Comment (SylviaWhittle) [reply utils/heightthresholding.py:65]: Oh I see, I did a find and replace which is why this was edited, not sure if it breaks anything so I wasn't sure what to do with it.

File: tests/test_filters.py [hunk 12]
```diff
@@ -14,29 +14,38 @@
 BASE_DIR = Path.cwd()
 RESOURCES = BASE_DIR / "tests" / "resources"
 
+# Example of how to add / update tests that rely on creating _.csv.bz2 files:
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/test_filters.py]: Done 👍

**Comment by SylviaWhittle** [comment]
> 

Fixed! 😄

File: tests/tracing/_regtest_outputs/test_dnatracing.test_tracestats.out (hunk 9 above)
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/tracing/_regtest_outputs/test_dnatracing.test_tracestats.out:5]: I have created a difference comparison of the binary grain masks that are passed to the flattening functions, I see only minor changes:

diff

File: tests/_regtest_outputs/test_run_topostats.test_process_scan_both.out (hunk 6 above)
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/_regtest_outputs/test_run_topostats.test_process_scan_both.out]: @ns-rse I updated the config used for the `process_scan` tests like you said in the `process_scan_config` fixture and the tests work well now while maintaining the correct values in `default_config.yaml` 

The lower thresholding capabilities are needed as they can be useful for other sample types.

File: utils/heightthresholding.py (hunk 7 above)
Comment (SylviaWhittle) [reply utils/heightthresholding.py:65]: I think I'll leave this in with the updated function name just in case it helps with clarity when refactoring

File: topostats/validation.py [hunk 13]
```diff
@@ -93,7 +95,8 @@ def validate_config(config: dict):
                     ),
                 ),
                 "otsu_threshold_multiplier": float,
-                "threshold_std_dev": lambda n: 0 < n <= 6,
+                "threshold_std_dev_lower": float,
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply topostats/validation.py:98]: Ok, that's not a problem if the Schema is to be updated.

Its slightly strange that anything above 10 standard deviations can be detected though as almost nothing will ever by 10 standard deviations from the mean. Central Limit Theorm posits that in roughly Gaussian distributions ~99.7% of observations are `mean -+ 3 * std` and so 10 times the standard deviation is going to include pretty much every observation and so setting the upper threshold to be that level and only detecting things larger than that would be highly unlikely to find anything.

I realise in scans of DNA and RNA the distribution is bi-modal and not Gaussian but this in turn makes the variance even higher. From memory the images of silicone I've seen are all "high" and there is perhaps likely to be lower variance but still 10 standard deviations from the mean should set the threshold so high that most points are excluded.

@SylviaWhittle introduction of histogram of heights might provide some insight here and it depends on what stage of processing is being considered.

Back on topic the specific test here should perhaps be (as @SylviaWhittle has suggested in chat)...

```suggestion
                "threshold_std_dev_lower": lambda n: n > 0
```

..and the same for `threshold_std_dev_upper` as we always want a positive value.

File: topostats/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ns-rse) [reply topostats/utils.py:208]: I kind of like dictionaries for this sort of thing as the simplify usage and keep things clear.

And so in `topostats/default_config.yaml` we can have...

```
threshold_std_dev:
  lower: 1.0
  upper: 1.0
```

As passing a dictionary(/list/tuple) reduces the number of options to a class/argument and avoids hitting the linting error [`too-many-arguments` / `R0913`](https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/too-many-arguments.html).

File: tests/tracing/_regtest_outputs/test_dnatracing.test_tracestats.out (hunk 9 above)
[the remark this answers is not in the document]
Comment (MaxGamill-Sheffield) [reply tests/tracing/_regtest_outputs/test_dnatracing.test_tracestats.out:5]: @SylviaWhittle So I've been looking into this and this is what I've found:
- The following plots show the differences between the image in the region around the second grain when processing with otsu. We see not much change:
Screenshot 2022-11-11 at 14 51 22

- This is reflected in the mask and difference masks:
Screenshot 2022-11-11 at 14 51 49
Screenshot 2022-11-11 at 14 52 05

- When we look at the traces that DNA tracing has made, we find that the disordered traces differ, the ordered traces become more aligned and the splined traces are not generated from the fitted trace. This means the difference in the fitted trace is causing the issue, and we can see it's down to the janky part of the fitted trace getting larger:
Screenshot 2022-11-11 at 14 59 44

- The slight increase in pixel heights around this area in the dev branch may be the reason for this:
Screenshot 2022-11-11 at 15 03 27

Therefor this might be one to tally up to poor molecule tracing and not a change of behaviour in flattening.

[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/tracing/_regtest_outputs/test_dnatracing.test_tracestats.out:5]: Thank you very much for this in-depth analysis @MaxGamill-Sheffield. I concur. The flattening is important to get right, and through histograms, experimentalist opinion and median plotting I am able to show that the new flattening is better, so I would say this could be left up to tweaks in the tracing functions to address.

Resolving the conversation, un-resolve if you don't agree 😄

**Comment by slackline** [comment]
Thanks for addressing the many and mostly minor changes @SylviaWhittle I think there is just this [one question](https://github.com/AFM-SPM/TopoStats/pull/351/files/046212e1fdee24791eff11371415977e5657e7a3#r1016737423) I had about what happens when a condition isn't met outstanding, its likely not essential as I can't think of why/how such a situation might arise (would a "scar" ever cause this?), but perhaps raising an exception or doing away with the `if` if its not required.

One thought is it was included because you encountered such problems during development. If so that might be informative/guide how it might occur and how to handle it.

File: topostats/filters.py [hunk 14]
```diff
@@ -86,90 +89,97 @@ def __init__(
...
+            LOGGER.info("median flattening with mask")
+        else:
+            read_matrix = image
+            LOGGER.info("median flattening without mask")
+
+        for j in range(image.shape[0]):
+            # Get the median of the row
+            m = np.nanmedian(read_matrix[j, :])
+            # print(m)
+            if not np.isnan(m):
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/filters.py:117]: @ns-rse 
Ah! Yes so this is an issue that occurs with images where a grain covers an entire row. That is to say that the data has a row that is completely `np.nan` after being masked. This happens when the image is particularly small and the grains are huge. The result of this is that the median becomes `np.nan` and therefore cannot be used arithmetically, hence, flattening just cannot occur for that file.

I will update this to raise an error, with a useful message. Do you have any suggestions as to what kind of error I should raise? I've not made a custom error before.

**Review by MaxGamill-Sheffield — CHANGES_REQUESTED** [review]
Hey Sylvia, great job on addressing things so quickly!

Only a few silly pedantic things. It won't let me add a comment to `plotting_dict` because it's technically unchanged but do you think we could:
Change in `filters` and `plotting_dict` image references to "initial_align" to be named "initial_median_flatten" instead? All other images reference the function used to generate the image in their name and this function is no longer called "initial_align" making it harder to find the function if there is a problem with the image (same for "masked_align").

File: tests/test_filters.py [hunk 15]
```diff
@@ -15,28 +15,14 @@
...
-    aligned_rows = test_filters_random.align_rows(test_filters_random.images["pixels"], mask=None)
+    median_flattened = test_filters_random.median_flatten(test_filters_random.images["pixels"], mask=None)
 
-    assert isinstance(aligned_rows, np.ndarray)
-    assert aligned_rows.shape == (1024, 1024)
-    np.testing.assert_allclose(aligned_rows, image_random_aligned_rows, **TOLERANCE)
+    assert isinstance(median_flattened, np.ndarray)
+    assert median_flattened.shape == (1024, 1024)
+
+    np.testing.assert_allclose(median_flattened, image_random_median_flattened, **TOLERANCE)
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply tests/test_filters.py:25]: I like the approach of summing the `1024x1024` arrays as it gives a non-graphical check that the images are _exactly_ the same which they should be each time.

File: tests/test_filters_minicircle.py [hunk 16]
```diff
@@ -47,16 +47,38 @@ def test_remove_x_y_tilt_unmasked(
...
+        **plotting_config,
+    ).plot_and_save()
+    return fig
+
+
 def test_get_threshold_otsu(minicircle_threshold_otsu: np.array) -> None:
     """Test calculation of threshold."""
     assert isinstance(minicircle_threshold_otsu.thresholds, dict)
-    assert minicircle_threshold_otsu.thresholds["upper"] == 28.58495414588038
+    assert minicircle_threshold_otsu.thresholds["upper"] == pytest.approx(-0.7471511148088736)
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply tests/test_filters_minicircle.py:73]: I wouldn't have expected it as this is `test_get_threshold_otsu()` and so there should only be one threshold, which by default is stored with the `upper` key and so `lower` shouldn't have an influence here I don't think.

File: tests/test_filters_minicircle.py [hunk 17]
```diff
@@ -8,17 +8,17 @@
 from topostats.plottingfuncs import Images
 
 # Specify the absolute and relattive tolerance for floating point comparison
-TOLERANCE = {"atol": 1e-07, "rtol": 1e-07}
+TOLERANCE = {"atol": 1e-06, "rtol": 1e-06}
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/test_filters_minicircle.py:11]: I changed these as for some reason whenever I run the tests, they always vary on that magnitude for that test. It doesn't happen for the other tests though...

File: tests/test_filters_minicircle.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/test_filters_minicircle.py:73]: I think it's because `minicircle_threshold_otsu` uses the `minicircle_initial_tilt_removal` fixture which has had processing applied to it. The methods of processing have changed and hence the image and therefore the thresholds have changed? Unless I'm missing something.

File: topostats/filters.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/filters.py:117]: Yeah @ns-rse discussed this a little in slack, should have posted it here. I'm in the process of writing a custom exception for it, though trying to find out if I can stop one image from processing while allowing the others to continue, then spitting an error out at the end.

File: topostats/filters.py [hunk 18]
```diff
@@ -86,90 +89,97 @@ def __init__(
...
-            2-D image to calculate row and column medians.
-        mask: np.ndarray
-            Boolean array of points to mask.
-        Returns
-        -------
-        dict
-            Dict of two Numpy arrays corresponding to row height value medians and column height value medians.
-        """
+    def remove_tilt(self, image: np.ndarray, mask: np.ndarray = None):
+        image = image.copy()
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/filters.py:139]: Added 👍

File: tests/test_filters_minicircle.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/test_filters_minicircle.py:11]: I am sure that it is just processors doing funky stuff with floats, and this culminating in a slightly higher uncertainty in calculation.

File: topostats/filters.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/filters.py]: Line of best fit. Added better comment 👍

File: topostats/plotting_dictionary.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/plotting_dictionary.yaml]: Done 👍

File: topostats/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/utils.py]: Added 👍

File: topostats/plottingfuncs.py [hunk 19]
```diff
@@ -96,6 +96,34 @@ def __init__(
...
+        -------
+        fig: plt.figure.Figure
+            Matplotlib.pyplot figure object
+        ax: plt.axes._subplots.AxesSubplot
+            Matplotlib.pyplot axes object
+        """
+
+        fig, ax = plt.subplots(1, 1, figsize=(8, 8))
+
+        ax.hist(self.data.flatten().astype(float), bins="auto", log=True)
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/plottingfuncs.py]: I could be wrong but I find that 'auto' provides a good amount of bins, whilst not causing the process to grind to a halt. I'm not sure what one would set this to to gain more insight into the data. More bins is better and 'auto' gives a good number in my opinion? (Having as many bins as different data points is not viable as it takes ages to plot, having 1e6-ish bins). Do say if you think I'm wrong on this though 😄 .

[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/plottingfuncs.py]: @ns-rse  With regards to the `edgecolor`, another instance of the tests doing things differently, in normal operation, the lines aren't there in the histogram plots. So in a quirky state of affairs, if the lines are there in the tests, then they won't be in the normal operation. Therefore as it stands, it behaves as intended but the test plots look worse.

image

File: tests/test_filters_minicircle.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (MaxGamill-Sheffield) [reply tests/test_filters_minicircle.py:73]: Makes sense to me that the initial image has changed (and this image from the previous `minicircle_initial_median_flatten` (the first in the sequence) which also zero-means the data hence the threshold is dramatically reduced.

File: topostats/filters.py [hunk 20]
```diff
@@ -86,129 +85,148 @@ def __init__(
...
             LOGGER.setLevel("ERROR")
 
-    def flatten_image(self, image: np.ndarray, mask: np.ndarray = None) -> np.ndarray:
-        """Flatten an image.
-
-        Flattening an image involves first aligining rows and then removing tilt, with a mask optionally applied to both
-        stages. These methods could be called independently but this method is provided as a convenience.
+    def median_flatten(self, image: np.ndarray, mask: np.ndarray = None, img_name: str = None) -> np.ndarray:
+        """
+        Uses the method of median differences to flatten the rows of an image, aligning the rows. Note this does not handle scars.
```
[the remark this answers is not in the document]
Comment (MaxGamill-Sheffield) [reply topostats/filters.py]: I've just ran the following file as an attempt to flatten the hertz noise Billie and Rob are seeing (https://drive.google.com/file/d/16QqZgNyHsdSoC4O-3sGhAZk_fTC5ImCZ/view?usp=share_link).

With this image I see that the scale bar changes between flattening steps, posing the question is there a common theme in this or do we still need a zero-average background step to ensure this has been done correctly?

02-initial_align_unmasked
03-initial_tilt_removal_unmasked
![Uploading 06-secondary_align_masked.png…]
08-quadratic_removal_masked

01-pixels

**Review by ns-rse — APPROVED** [review]
Happy to approve this, thanks for all your hard work on this @SylviaWhittle 

I think @MaxGamill-Sheffield question about whether the zero-averaged step is required should be treated as a separate issue/another iteration as this PR is very large and there is the desire to switch branches before Wednesday.

**Review by MaxGamill-Sheffield — APPROVED** [review]
Looks good to meee

**Comment by SylviaWhittle** [comment]
Merging and will probably add the custom bin numbers in a future PR as it's not strictly needed here and I am honestly done with the size of this PR 😆
